design/design-generic-references.md

```diff
@@ -0,0 +1,477 @@
+
+```go
+type Reference struct {
+// Name of the referenced object.
+Name string `json:"name"`
+
+// Policies for referencing.
+// +optional
+Policy *Policy `json:"policy,omitempty"`
+}
+```
+
+### GenericReference Type
+
+We would like to introduce `GenericReference` field type:
+
+```go
+type GenericReference struct {
+  Reference
+  
+  // ApiVersion of the referenced object.
+  ApiVersion string `json:"apiVersion"`
+  
+  // Kind of the referenced object.
+  Kind string `json:"kind"`
+  
+  // FieldPath of the value within the referenced object.
+  FieldPath string `json:"fieldPath"`
+}
+```
+
+The syntax of the `fieldPath` field is the one used in Composition, in line with
+the [Kubernetes API Conventions on field selection].
+
+Reference fields of the above type describe completely the location of 
+the value that is about to be set on the counterpart field. In case of 
+`datasync.aws.upbound.io` `Task`, an instance might look like:
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnRef:
+      name: source-location
+      apiVersion: datasync.aws.upbound.io/v1beta1
+      kind: LocationS3
+      fieldPath: metadata.annotations[crossplane.io/external-name]
+```
+
+Such approach lets a user to point the `sourceLocationArnRef` field to a different
+resource at the runtime.
+
+### GenericSelector Type
+
+Similar to `GenericReference`, we are going to introduce `GenericSelector` type:
+
+```go
+type GenericSelector struct {
+  // MatchLabels ensures an object with matching labels is selected.
+  MatchLabels map[string]string `json:"matchLabels,omitempty"`
+  
+  // Policies for selection.
+  // +optional
+  Policy *Policy `json:"policy,omitempty"`
+  
+  // ApiVersion of the referenced object.
+  ApiVersion string `json:"apiVersion"`
+  
+  // Kind of the referenced object.
+  Kind string `json:"kind"`
+  
+  // FieldPath of the value within the referenced object.
+  FieldPath string `json:"fieldPath"`
+}
+```
+
+so that one can reference the external object when its name is
+unknown/irrelevant:
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnSelector:
+      matchLabels:
+        foo: bar
+      apiVersion: datasync.aws.upbound.io/v1beta1
+      kind: LocationS3
+      fieldPath: metadata.annotations[crossplane.io/external-name]
+```
+
+Although the proposed solution sounds like a natural evolution of the existing
+reference mechanism, its adoption would require that:
+
+* Providers implement the support for it.
+* RBAC permissions need to be relaxed, if provider `A` needs to refer values from
+  instances owned by provider `B`. In case of provider families, each provider has
+  enough rights to access all resources within the family, requiring no changes
```

Comment:
I wonder if enterprise users may ask for fine-granular access policies in the future. Currently, access is mutually granted to the (service accounts of the) providers A & B, if A & B are in the same family, even if their MRs are not supposed to refer to each other. This's been a natural evolution from the monoliths.